File: src/main/java/com/example/dalattravel/service/PDPTWService.java
package com.example.dalattravel.service;

import com.example.dalattravel.dto.PDPTWNode;
import com.example.dalattravel.dto.PDPTWRoute;
import com.example.dalattravel.model.Passenger;
import com.example.dalattravel.model.Vehicle;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.*;

@Service
public class PDPTWService {

    public PDPTWRoute optimizeRoute(Vehicle vehicle, List<Passenger> passengers) {
        if (passengers == null || passengers.isEmpty()) {
            PDPTWRoute route = new PDPTWRoute();
            route.setNodes(new ArrayList<>());
            route.setTotalDistance(0);
            route.setStartTime(vehicle.getDepartureTime());
            route.setEndTime(vehicle.getDepartureTime());
            route.setTotalCost(BigDecimal.ZERO);
            return route;
        }

        List<PDPTWNode> nodes = new ArrayList<>();

        for (Passenger passenger : passengers) {
            PDPTWNode pickup = new PDPTWNode();
            pickup.setLatitude(passenger.getPickupLatitude());
            pickup.setLongitude(passenger.getPickupLongitude());
            pickup.setAddress(passenger.getPickupAddress() != null ? passenger.getPickupAddress() : "");
            pickup.setType("pickup");
            pickup.setPassengerId(passenger.getId());
            pickup.setEarliestTime(passenger.getPreferredDepartureTime().minusMinutes(30));
            pickup.setLatestTime(passenger.getPreferredDepartureTime().plusHours(2));
            pickup.setServiceTime(5);
            nodes.add(pickup);

            PDPTWNode dropoff = new PDPTWNode();
            dropoff.setLatitude(passenger.getDropoffLatitude());
            dropoff.setLongitude(passenger.getDropoffLongitude());
            dropoff.setAddress(passenger.getDropoffAddress() != null ? passenger.getDropoffAddress() : "");
            dropoff.setType("dropoff");
            dropoff.setPassengerId(passenger.getId());
            dropoff.setEarliestTime(passenger.getPreferredArrivalTime() != null ? 
                    passenger.getPreferredArrivalTime().minusMinutes(30) : passenger.getPreferredDepartureTime().plusHours(1));
            dropoff.setLatestTime(passenger.getPreferredArrivalTime() != null ? 
                    passenger.getPreferredArrivalTime().plusHours(2) : passenger.getPreferredDepartureTime().plusHours(4));
            dropoff.setServiceTime(5);
            nodes.add(dropoff);
        }

        PDPTWRoute route = new PDPTWRoute();
        route.setNodes(new ArrayList<>());
        route.setStartTime(vehicle.getDepartureTime());

        List<PDPTWNode> unvisited = new ArrayList<>(nodes);
        LocalDateTime currentTime = vehicle.getDepartureTime();
        double currentLat = vehicle.getPickupLatitude();
        double currentLng = vehicle.getPickupLongitude();
        Set<Integer> pickedUpPassengers = new HashSet<>();

        while (!unvisited.isEmpty()) {
            PDPTWNode nextNode = null;
            double minDistance = Double.MAX_VALUE;

            for (PDPTWNode node : unvisited) {
                if ("dropoff".equals(node.getType()) && !pickedUpPassengers.contains(node.getPassengerId()))
                    continue;

                if (currentTime.isAfter(node.getLatestTime()))
                    continue;

                double distance = calculateDistance(currentLat, currentLng, node.getLatitude(), node.getLongitude());
                double tempTravelTimeMinutes = (distance / 50.0) * 60;

                LocalDateTime tempArrivalTime = currentTime.plusMinutes((long) tempTravelTimeMinutes);
                if (tempArrivalTime.isBefore(node.getEarliestTime()))
                    tempArrivalTime = node.getEarliestTime();

                if (!tempArrivalTime.isAfter(node.getLatestTime())) {
                    double score = distance;
                    if (tempArrivalTime.isBefore(node.getEarliestTime()))
                        score += 1000;

                    if (score < minDistance) {
                        minDistance = score;
                        nextNode = node;
                    }
                }
            }

            if (nextNode == null) {
                final double cLat = currentLat;
                final double cLng = currentLng;
                nextNode = unvisited.stream()
                        .min(Comparator.comparingDouble(n -> calculateDistance(cLat, cLng, n.getLatitude(), n.getLongitude())))
                        .orElse(null);

                if (nextNode == null) break;
            }

            double distanceToNode = calculateDistance(currentLat, currentLng, nextNode.getLatitude(), nextNode.getLongitude());
            double travelTime = (distanceToNode / 50.0) * 60;
            LocalDateTime arrivalTime = currentTime.plusMinutes((long) travelTime);

            if (arrivalTime.isBefore(nextNode.getEarliestTime()))
                arrivalTime = nextNode.getEarliestTime();

            currentTime = arrivalTime.plusMinutes(nextNode.getServiceTime());
            currentLat = nextNode.getLatitude();
            currentLng = nextNode.getLongitude();

            route.getNodes().add(nextNode);
            unvisited.remove(nextNode);

            if ("pickup".equals(nextNode.getType()))
                pickedUpPassengers.add(nextNode.getPassengerId());
            else if ("dropoff".equals(nextNode.getType()))
                pickedUpPassengers.remove(nextNode.getPassengerId());
        }

        route.setTotalDistance(0);
        double routeLat = vehicle.getPickupLatitude();
        double routeLng = vehicle.getPickupLongitude();

        for (PDPTWNode node : route.getNodes()) {
            route.setTotalDistance(route.getTotalDistance() + calculateDistance(routeLat, routeLng, node.getLatitude(), node.getLongitude()));
            routeLat = node.getLatitude();
            routeLng = node.getLongitude();
        }

        route.setEndTime(currentTime);
        BigDecimal costPerKm = vehicle.getCostPerKm() != null ? vehicle.getCostPerKm() : BigDecimal.ZERO;
        route.setTotalCost(BigDecimal.valueOf(route.getTotalDistance()).multiply(costPerKm));

        return route;
    }

    private double calculateDistance(double lat1, double lon1, double lat2, double lon2) {
        final double R = 6371;
        double dLat = Math.toRadians(lat2 - lat1);
        double dLon = Math.toRadians(lon2 - lon1);
        double a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
                   Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2)) *
                   Math.sin(dLon / 2) * Math.sin(dLon / 2);
        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
        return R * c;
    }
}

File: DoAnChuyenNganh-main/DoAnCoSo/WebDuLichDaLat/Services/PDPTWService.cs
using WebDuLichDaLat.Models;
using System.Collections.Generic;
using System.Linq;

namespace WebDuLichDaLat.Services
{
    /// <summary>
    /// Dich vu giai quyet bai toan Pickup and Delivery Problem with Time Windows (PDPTW).
    /// Giup toi uu hoa thu tu don va tra khach dua tren toa do dia ly va thoi gian yeu cau.
    /// </summary>
    public class PDPTWService
    {
        /// <summary>
        /// Dai dien cho mot diem dung (don hoac tra khach) trong lo trinh.
        /// </summary>
        public class PDPTWNode
        {
            public double Latitude { get; set; }
            public double Longitude { get; set; }
            public string Address { get; set; }
            public string Type { get; set; } // Gia tri: "pickup" hoac "dropoff"
            public int? PassengerId { get; set; }
            public DateTime EarliestTime { get; set; }
            public DateTime LatestTime { get; set; }
            public int ServiceTime { get; set; } = 5; // Thoi gian dung cho moi diem (phut)
        }

        /// <summary>
        /// Ket qua lo trinh sau khi da duoc toi uu hoa.
        /// </summary>
        public class PDPTWRoute
        {
            public List<PDPTWNode> Nodes { get; set; } = new List<PDPTWNode>();
            public double TotalDistance { get; set; }
            public DateTime StartTime { get; set; }
            public DateTime EndTime { get; set; }
            public decimal TotalCost { get; set; }
        }

        /// <summary>
        /// Thuc hien toi uu hoa lo trinh di chuyen cho phuong tien va danh sach hanh khach.
        /// </summary>
        /// <param name="vehicle">Thong tin phuong tien van chuyen</param>
        /// <param name="passengers">Danh sach hanh khach can don/tra</param>
        /// <returns>Doi tuong PDPTWRoute chua lo trinh toi uu</returns>
        public PDPTWRoute OptimizeRoute(Vehicle vehicle, List<Passenger> passengers)
        {
            if (!passengers.Any())
            {
                return new PDPTWRoute
                {
                    Nodes = new List<PDPTWNode>(),
                    TotalDistance = 0,
                    StartTime = vehicle.DepartureTime,
                    EndTime = vehicle.DepartureTime,
                    TotalCost = 0
                };
            }

            // Khoi tao danh sach cac diem nut (Nodes) cho viec don va tra khach
            var nodes = new List<PDPTWNode>();

            foreach (var passenger in passengers)
            {
                nodes.Add(new PDPTWNode
                {
                    Latitude = passenger.PickupLatitude,
                    Longitude = passenger.PickupLongitude,
                    Address = passenger.PickupAddress ?? "",
                    Type = "pickup",
                    PassengerId = passenger.Id,
                    EarliestTime = passenger.PreferredDepartureTime.AddMinutes(-30),
                    LatestTime = passenger.PreferredDepartureTime.AddHours(2),
                    ServiceTime = 5
                });

                nodes.Add(new PDPTWNode
                {
                    Latitude = passenger.DropoffLatitude,
                    Longitude = passenger.DropoffLongitude,
                    Address = passenger.DropoffAddress ?? "",
                    Type = "dropoff",
                    PassengerId = passenger.Id,
                    EarliestTime = passenger.PreferredArrivalTime?.AddMinutes(-30) ?? passenger.PreferredDepartureTime.AddHours(1),
                    LatestTime = passenger.PreferredArrivalTime?.AddHours(2) ?? passenger.PreferredDepartureTime.AddHours(4),
                    ServiceTime = 5
                });
            }

            // Su dung thuat toan Greedy (Nearest Neighbor) ket hop voi cac rang buoc ve thoi gian
            var route = new PDPTWRoute
            {
                Nodes = new List<PDPTWNode>(),
                StartTime = vehicle.DepartureTime
            };

            var unvisited = new List<PDPTWNode>(nodes);
            var currentTime = vehicle.DepartureTime;
            double currentLat = vehicle.PickupLatitude;
            double currentLng = vehicle.PickupLongitude;
            var pickedUpPassengers = new HashSet<int>();

            while (unvisited.Any())
            {
                PDPTWNode nextNode = null;
                double minDistance = double.MaxValue;

                foreach (var node in unvisited)
                {
                    // Rang buoc nghiep vu: Hanh khach phai duoc don (pickup) truoc khi tra (dropoff)
                    if (node.Type == "dropoff" && !pickedUpPassengers.Contains(node.PassengerId.Value))
                        continue;

                    // Rang buoc ve cua so thoi gian (Time Window)
                    if (currentTime > node.LatestTime)
                        continue;

                    double distance = CalculateDistance(currentLat, currentLng, node.Latitude, node.Longitude);
                    double tempTravelTime = distance / 50.0 * 60; // Gia thiet van toc trung binh la 50km/h

                    DateTime tempArrivalTime = currentTime.AddMinutes(tempTravelTime);
                    if (tempArrivalTime < node.EarliestTime)
                        tempArrivalTime = node.EarliestTime;

                    if (tempArrivalTime <= node.LatestTime)
                    {
                        // Lua chon diem dung tiep theo dua tren tieu chi khoang cach va thoi gian hop le
                        double score = distance;
                        if (tempArrivalTime < node.EarliestTime)
                            score += 1000; // Ap dung hinh phat (Penalty) neu den qua som so voi cua so thoi gian

                        if (score < minDistance)
                        {
                            minDistance = score;
                            nextNode = node;
                        }
                    }
                }

                if (nextNode == null)
                {
                    // Truong hop khong tim thay diem dung hop le theo thoi gian, lua chon diem gan nhat con lai
                    nextNode = unvisited
                        .OrderBy(n => CalculateDistance(currentLat, currentLng, n.Latitude, n.Longitude))
                        .FirstOrDefault();

                    if (nextNode == null)
                        break;
                }

                // Cap nhat thong tin trang thai lo trinh
                double distanceToNode = CalculateDistance(currentLat, currentLng, nextNode.Latitude, nextNode.Longitude);
                double travelTime = distanceToNode / 50.0 * 60;
                DateTime arrivalTime = currentTime.AddMinutes(travelTime);

                if (arrivalTime < nextNode.EarliestTime)
                    arrivalTime = nextNode.EarliestTime;

                currentTime = arrivalTime.AddMinutes(nextNode.ServiceTime);
                currentLat = nextNode.Latitude;
                currentLng = nextNode.Longitude;

                route.Nodes.Add(nextNode);
                unvisited.Remove(nextNode);

                if (nextNode.Type == "pickup")
                    pickedUpPassengers.Add(nextNode.PassengerId.Value);
                else if (nextNode.Type == "dropoff")
                    pickedUpPassengers.Remove(nextNode.PassengerId.Value);
            }

            // Tong hop ket qua lo trinh (Tong khoang cach va chi phi)
            route.TotalDistance = 0;
            double routeLat = vehicle.PickupLatitude;
            double routeLng = vehicle.PickupLongitude;

            foreach (var node in route.Nodes)
            {
                route.TotalDistance += CalculateDistance(routeLat, routeLng, node.Latitude, node.Longitude);
                routeLat = node.Latitude;
                routeLng = node.Longitude;
            }

            route.EndTime = currentTime;
            route.TotalCost = (decimal)route.TotalDistance * vehicle.CostPerKm;

            return route;
        }

        private double CalculateDistance(double lat1, double lon1, double lat2, double lon2)
        {
            const double R = 6371; // Ban kinh Trai Dat (km)
            double dLat = ToRadians(lat2 - lat1);
            double dLon = ToRadians(lon2 - lon1);
            double a = Math.Sin(dLat / 2) * Math.Sin(dLat / 2) +
                       Math.Cos(ToRadians(lat1)) * Math.Cos(ToRadians(lat2)) *
                       Math.Sin(dLon / 2) * Math.Sin(dLon / 2);
            double c = 2 * Math.Atan2(Math.Sqrt(a), Math.Sqrt(1 - a));
            return R * c;
        }

        private double ToRadians(double angle) => angle * Math.PI / 180;
    }
}






